presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java

```diff
@@ -0,0 +1,364 @@
+            node = removeLimitNode(node);
+        }
+
+        node = ensureJoinSymbolsAreReturned(node, correlatedPredicates);
+
+        return decorrelatedNode(correlatedPredicates, node, correlation);
+    }
+
+    private static Optional<DecorrelatedNode> decorrelatedNode(
+            List<Expression> correlatedPredicates,
+            PlanNode node,
+            List<Symbol> correlation)
+    {
+        if (DependencyExtractor.extractUnique(node).stream().anyMatch(correlation::contains)) {
+            // node is still correlated ; /
+            return Optional.empty();
+        }
+        return Optional.of(new DecorrelatedNode(correlatedPredicates, node));
+    }
+
+    private static Predicate<Expression> isUsingPredicate(List<Symbol> symbols)
+    {
+        return expression -> symbols.stream().anyMatch(DependencyExtractor.extractUnique(expression)::contains);
+    }
+
+    private PlanNode updateFilterNode(PlanNodeSearcher filterNodeSearcher, List<Expression> newPredicates)
+    {
+        if (newPredicates.isEmpty()) {
+            return filterNodeSearcher.removeAll();
+        }
+        FilterNode oldFilterNode = Iterables.getOnlyElement(filterNodeSearcher.findAll());
+        FilterNode newFilterNode = new FilterNode(
+                idAllocator.getNextId(),
+                oldFilterNode.getSource(),
+                ExpressionUtils.combineConjuncts(newPredicates));
+        return filterNodeSearcher.replaceAll(newFilterNode);
+    }
+
+    private static PlanNode removeLimitNode(PlanNode node)
+    {
+        node = searchFrom(node)
+                .where(LimitNode.class::isInstance)
+                .skipOnlyWhen(ProjectNode.class::isInstance)
+                .removeFirst();
+        return node;
+    }
+
+    private PlanNode ensureJoinSymbolsAreReturned(PlanNode scalarAggregationSource, List<Expression> joinPredicate)
+    {
+        Set<Symbol> joinExpressionSymbols = DependencyExtractor.extractUnique(joinPredicate);
+        ExtendProjectionRewriter extendProjectionRewriter = new ExtendProjectionRewriter(
+                idAllocator,
+                joinExpressionSymbols);
+        return rewriteWith(extendProjectionRewriter, scalarAggregationSource);
+    }
+
+    private static boolean isSupportedPredicate(Expression predicate)
+    {
+        AtomicBoolean isSupported = new AtomicBoolean(true);
+        new DefaultTraversalVisitor<Void, AtomicBoolean>()
+        {
+            @Override
+            protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, AtomicBoolean context)
+            {
+                if (node.getType() != LogicalBinaryExpression.Type.AND) {
+                    context.set(false);
+                }
+                return null;
+            }
+        }.process(predicate, isSupported);
+        return isSupported.get();
+    }
+
+    private static class DecorrelatedNode
```

Comment:
https://github.com/prestodb/presto/pull/7888/commits/690894b11641c445aebc41c4a389bf83c5e31a08#diff-1a430cab232220a3d693ed1d27066e9fR404 ... Do you see potential for re-use?